[Blaze i4] Setting local notification for no campaign during 30 days

irfano opened the pull request:
<!-- Remember about a good descriptive title. -->
<!-- Id number of the GitHub issue this PR addresses. -->
Part of https://github.com/woocommerce/woocommerce-android/issues/12290

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This introduces a new local notification which is a reminder for Blaze. This PR only covers scheduling and displaying the notification. The logic determining when to show the notification will be addressed in the following PR. Tapping the notification should open the campaign creation screen.

### Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->
This PR does not trigger the notification. To trigger the notification 31 seconds after launching the app, apply this patch (ensure that notification permissions are enabled): [Trigger_blaze_reminder_notification_for_testing.patch](https://github.com/user-attachments/files/16624484/Trigger_blaze_reminder_notification_for_testing.patch)

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
<img src="https://github.com/user-attachments/assets/8d9f3767-797c-434d-baaf-aca1c5c781a7" width=300>

- [x] I have considered adding unit tests for this change. If I decided not to add them, I have provided a brief explanation below (optional):
- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

irfano committed:
Optimize LocalNotification functions

irfano committed:
Add BLAZE_NO_CAMPAIGN_REMINDER notification type

irfano committed:
Update PreconditionCheckWorker logic

The `proceedIfFreeTrialAndMatchesSite()` function was left over from an unused feature.

irfano committed:
Open the campaign list when the Blaze reminder notification is tapped

irfano committed:
Add BLAZE_NO_CAMPAIGN_REMINDER_SHOWN key to AppPrefs

irfano committed:
Set blaze reminder notification as shown when tapped

irfano committed:
Fix typo in BlazeCampaignPaymentSummaryScreen

irfano committed:
Merge branch 'trunk' into issue/12290-blaze-reminder-notification

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/notifications/local/LocalNotification.kt:
```diff
@@ -15,9 +16,19 @@ sealed class LocalNotification(
     val id = type.hashCode()
 
-    abstract fun getDescriptionString(resourceProvider: ResourceProvider): String
+    open fun getTitleString(resourceProvider: ResourceProvider) = resourceProvider.getString(title)
```
Minor np, but do we need these to be `open` ? I know this was set that way before, but it seems that with the current usage we don't need them to be `open` yet and can probably default the 2 functions to be final.
Wdyt?

irfano replied:
Good catch! I removed `open`s: 4eb9358961ad30e48b2b7684bf4c457ff357ef4e

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/notifications/local/PreconditionCheckWorker.kt:
```diff
@@ -32,13 +31,13 @@ class PreconditionCheckWorker @AssistedInject constructor(
     }
 
-    private fun proceedIfFreeTrialAndMatchesSite(siteId: Long): Result {
+    private fun proceedIfValidSite(siteId: Long): Result {
```
Thanks for catching this piece of legacy naming 👍🏼

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivityViewModel.kt:
```diff
@@ -260,6 +262,11 @@ class MainActivityViewModel @Inject constructor(
             )
+            LocalNotificationType.fromString(notification.tag)?.let {
+                when (it) {
+                    BLAZE_NO_CAMPAIGN_REMINDER -> triggerEvent(ShowAllCampaigns)
```
Hmm, unsure if you plan on changing this in subsequent PRs. But haven't seen any updates in the [next PR](https://github.com/woocommerce/woocommerce-android/pull/12374/files). Why open the campaign list screen? Afaik we should open the campaign creation flow: pe5sF9-331-p2

irfano replied:
Good catch again! Somehow, I thought the campaign list was the campaign creation screen. I've updated it: c30f1d7082e84af944bcaf7325c34555822f2467.

JorgeMucientes reviewed (approved):
Hey @irfano great work! 

I left a couple comments. One of them worth checking before merging eventhough it can be addressed in the [next PR](https://github.com/woocommerce/woocommerce-android/pull/12374). So I'll pre approve the PR but I think we need to update the destination when tapping on the PR. 

Also, I'd like to align with iOS on when they plan on releasing this local reminders. Becasuse we might want to align and release at the same time. In which case we'll need to add feature flag to the notification scheduling.

irfano committed:
Remove open from functions in LocalNotification

irfano committed:
Update the destination of the Blaze reminder notification

Previously, tapping the notification was opening the campaign list screen. Now, it will open the campaign creation screen.

irfano commented:
Thanks for catching those, @JorgeMucientes. I requested another review because the changes I added were not minor.

> Also, I'd like to align with iOS on when they plan on releasing this local reminders. Becasuse we might want to align and release at the same time. In which case we'll need to add feature flag to the notification scheduling.

Thanks for the reminder. I think we can at least merge this one since this PR doesn’t schedule the notification.

JorgeMucientes reviewed (approved):
Thanks for tackling the changes @irfano. Great job!! Code looks good and works as expected :shipit:
